### PR TITLE
 Add a variant using Ractors

### PR DESCRIPTION
**Pull request by eregon**
It seems to work correctly, but frequently fails with N >= 4:
```
time ruby 3-sequential-lee.rb inputs/testBoard.txt
routes: 203
cost:   3304
depth:  3
ruby 3-sequential-lee.rb inputs/testBoard.txt  1.04s user 0.01s system 99% cpu 1.051 total

time N=1 ruby 7-ractor-tvar-lee.rb inputs/testBoard.txt       
<internal:ractor>:38: warning: Ractor is experimental, and the behavior may change in future versions of Ruby! Also there are many implementation issues.
routes:      203
committed:   203
aborted:     0
cost:        3304
depth:       3
N=1 ruby 7-ractor-tvar-lee.rb inputs/testBoard.txt  1.68s user 0.06s system 106% cpu 1.644 total

time N=2 ruby 7-ractor-tvar-lee.rb inputs/testBoard.txt
<internal:ractor>:38: warning: Ractor is experimental, and the behavior may change in future versions of Ruby! Also there are many implementation issues.
routes:      203
committed:   203
aborted:     66
cost:        3299
depth:       3
N=2 ruby 7-ractor-tvar-lee.rb inputs/testBoard.txt  5.63s user 2.05s system 188% cpu 4.078 total

time N=3 ruby 7-ractor-tvar-lee.rb inputs/testBoard.txt
<internal:ractor>:38: warning: Ractor is experimental, and the behavior may change in future versions of Ruby! Also there are many implementation issues.
routes:      203
committed:   203
aborted:     115
cost:        3291
depth:       3
N=3 ruby 7-ractor-tvar-lee.rb inputs/testBoard.txt  11.05s user 7.06s system 245% cpu 7.379 total

time N=4 ruby 7-ractor-tvar-lee.rb inputs/testBoard.txt
<internal:ractor>:38: warning: Ractor is experimental, and the behavior may change in future versions of Ruby! Also there are many implementation issues.
routes:      203
committed:   203
aborted:     175
cost:        3302
depth:       3
N=4 ruby 7-ractor-tvar-lee.rb inputs/testBoard.txt  15.36s user 11.31s system 310% cpu 8.576 total

time N=4 ruby 7-ractor-tvar-lee.rb inputs/testBoard.txt
<internal:ractor>:38: warning: Ractor is experimental, and the behavior may change in future versions of Ruby! Also there are many implementation issues.
#<Thread:0x0000561185ba7ec0 run> terminated with exception (report_on_exception is true):
/home/eregon/code/ruby-stm-lee-demo/lib/lee/adjacent.rb:5:in `point_on_board?': undefined method `x' for 0...75:Range (NoMethodError)
	from /home/eregon/code/ruby-stm-lee-demo/lib/lee/adjacent.rb:25:in `block in adjacent'
	from /home/eregon/code/ruby-stm-lee-demo/lib/lee/adjacent.rb:25:in `select'
<internal:ractor>:130:in `take': thrown by remote Ractor. (Ractor::RemoteError)
	from 7-ractor-tvar-lee.rb:167:in `each'
	from 7-ractor-tvar-lee.rb:167:in `<main>'
/home/eregon/code/ruby-stm-lee-demo/lib/lee/adjacent.rb:5:in `point_on_board?': undefined method `x' for 0...75:Range (NoMethodError)
	from /home/eregon/code/ruby-stm-lee-demo/lib/lee/adjacent.rb:25:in `block in adjacent'
	from /home/eregon/code/ruby-stm-lee-demo/lib/lee/adjacent.rb:25:in `select'
	from /home/eregon/code/ruby-stm-lee-demo/lib/lee/adjacent.rb:25:in `adjacent'
	from 7-ractor-tvar-lee.rb:42:in `block (2 levels) in expand'
	from 7-ractor-tvar-lee.rb:40:in `each'
	from 7-ractor-tvar-lee.rb:40:in `block in expand'
	from 7-ractor-tvar-lee.rb:37:in `loop'
	from 7-ractor-tvar-lee.rb:37:in `expand'
	from 7-ractor-tvar-lee.rb:153:in `block (4 levels) in <main>'
	from <internal:thread>:25:in `atomically'
	from 7-ractor-tvar-lee.rb:152:in `block (3 levels) in <main>'
	from 7-ractor-tvar-lee.rb:149:in `loop'
	from 7-ractor-tvar-lee.rb:149:in `block (2 levels) in <main>'
zsh: exit 1     N=4 ruby 7-ractor-tvar-lee.rb inputs/testBoard.txt
N=4 ruby 7-ractor-tvar-lee.rb inputs/testBoard.txt  8.50s user 5.32s system 323% cpu 4.273 total

Other, similar error:
/home/eregon/code/ruby-stm-lee-demo/lib/lee/adjacent.rb:5:in `point_on_board?': undefined method `x' for 0...75:Range (NoMethodError)
	from /home/eregon/code/ruby-stm-lee-demo/lib/lee/adjacent.rb:25:in `block in adjacent'
	from /home/eregon/code/ruby-stm-lee-demo/lib/lee/adjacent.rb:25:in `select'
	from /home/eregon/code/ruby-stm-lee-demo/lib/lee/adjacent.rb:25:in `adjacent'
	from 7-ractor-tvar-lee.rb:42:in `block (2 levels) in expand'
	from 7-ractor-tvar-lee.rb:40:in `each'
	from 7-ractor-tvar-lee.rb:40:in `block in expand'
	from 7-ractor-tvar-lee.rb:37:in `loop'
	from 7-ractor-tvar-lee.rb:37:in `expand'
	from 7-ractor-tvar-lee.rb:153:in `block (4 levels) in <main>'
	from <internal:thread>:25:in `atomically'
	from 7-ractor-tvar-lee.rb:152:in `block (3 levels) in <main>'
	from 7-ractor-tvar-lee.rb:149:in `loop'
	from 7-ractor-tvar-lee.rb:149:in `block (2 levels) in <main>'
```

Which seems a clear bug of Ractor/CRuby, because `point` should be a Point, not an Integer there.

### CODE REVIEW

**Comment by eregon** [comment]
The diff compared to `6-concurrent-tvar-lee.rb` is relatively small overall: https://gist.github.com/eregon/a02e251b7f0412b8191c3effef894f37

File: 7-ractor-tvar-lee.rb [hunk 1]
```diff
@@ -0,0 +1,180 @@
+# Solves a board using TVar actually on two uncontrolled concurrent threads.
```
Comment (chrisseaton) [inline 7-ractor-tvar-lee.rb]: Can you update this?

File: 7-ractor-tvar-lee.rb [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            |worklist, board, obstructed, depth, solutions_ractor, expansions_dir|
+    loop do
+      route = worklist.take
+
+      Thread.atomically do
+        cost = expand(board, obstructed, depth, route)
+        solution = solve(board, route, cost)
+
+        if expansions_dir
+        Lee.draw board, solutions.values, [[cost.keys, solution]], File.join(expansions_dir, "expansion-#{route.object_id}.svg")
```
Comment (chrisseaton) [inline 7-ractor-tvar-lee.rb]: Indentation.

Comment (eregon) [reply 7-ractor-tvar-lee.rb]: Right, it's the same in other files, should I fix it there too?

Comment (chrisseaton) [reply 7-ractor-tvar-lee.rb]: Ah sorry yes please.

Comment (eregon) [reply 7-ractor-tvar-lee.rb]: Done

File: 7-ractor-tvar-lee.rb (hunk 1 above)
Comment (eregon) [reply 7-ractor-tvar-lee.rb]: Done

**Review by chrisseaton — APPROVED** [review]

**Comment by chrisseaton** [comment]
Finished in 9d7b6b2d7a9e491858f31c5761882d8e8ba744ec.

**Pull request merged**